tket2-py/src/circuit/cost.rs

```diff
@@ -0,0 +1,168 @@
+//!
+
+use std::cmp::Ordering;
+use std::iter::Sum;
+use std::ops::{Add, AddAssign, Sub};
+
+use pyo3::{prelude::*, PyTypeInfo};
+use tket2::circuit::cost::{CircuitCost, CostDelta};
+
+/// A generic circuit cost, backed by an arbitrary python object.
+#[pyclass]
+#[derive(Clone, Debug)]
+#[pyo3(name = "CircuitCost")]
+pub struct PyCircuitCost {
+    /// Generic python cost object.
+    pub cost: PyObject,
+}
+
+#[pymethods]
+impl PyCircuitCost {
+    /// Create a new circuit cost.
+    #[new]
+    pub fn new(cost: PyObject) -> Self {
+        Self { cost }
+    }
+}
+
+impl Default for PyCircuitCost {
+    fn default() -> Self {
+        Python::with_gil(|py| PyCircuitCost { cost: py.None() })
+    }
+}
+
+impl Add for PyCircuitCost {
+    type Output = PyCircuitCost;
+
+    fn add(self, rhs: PyCircuitCost) -> Self::Output {
+        Python::with_gil(|py| {
+            let cost = self
+                .cost
+                .call_method1(py, "__add__", (rhs.cost,))
+                .expect("Could not add circuit cost objects.");
+            PyCircuitCost { cost }
+        })
+    }
+}
+
+impl AddAssign for PyCircuitCost {
+    fn add_assign(&mut self, rhs: Self) {
+        Python::with_gil(|py| {
+            let cost = self
+                .cost
+                .call_method1(py, "__add__", (rhs.cost,))
+                .expect("Could not add circuit cost objects.");
+            self.cost = cost;
+        })
+    }
+}
+
+impl Sub for PyCircuitCost {
+    type Output = PyCircuitCost;
+
+    fn sub(self, rhs: PyCircuitCost) -> Self::Output {
+        Python::with_gil(|py| {
+            let cost = self
+                .cost
+                .call_method1(py, "__sub__", (rhs.cost,))
+                .expect("Could not subtract circuit cost objects.");
+            PyCircuitCost { cost }
+        })
+    }
+}
+
+impl Sum for PyCircuitCost {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        Python::with_gil(|py| {
+            let mut acc = None;
+            for c in iter {
+                match &mut acc {
+                    None => acc = Some(c.cost),
+                    Some(cost) => {
+                        *cost = cost
+                            .call_method1(py, "__add__", (c.cost,))
+                            .expect("Could not add circuit cost objects.")
+                    }
+                }
+            }
+            PyCircuitCost {
+                cost: acc.unwrap_or_else(|| py.None()),
```

Comment:
```suggestion
                cost: acc.unwrap_or_else(py.None),
```

Reply:
Might be making things less clear, feel free to drop

Reply:
`py` is a variable, so it wouldn't work

Reply:
Ah! Sorry :-!